src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py

```diff
@@ -0,0 +1,173 @@
+
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Iterator
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+
+from frequenz.channels import Broadcast, Receiver
+from frequenz.channels.util import Merge
+
+from frequenz.sdk import microgrid
+from frequenz.sdk._internal.asyncio import cancel_and_await
+
+from ...microgrid.component import (
+    EVChargerCableState,
+    EVChargerComponentState,
+    EVChargerData,
+)
+
+
+class EVChargerState(Enum):
+    """State of individual ev charger."""
+
+    UNSPECIFIED = "UNSPECIFIED"
+    IDLE = "IDLE"
+    EV_PLUGGED = "EV_PLUGGED"
+    EV_LOCKED = "EV_LOCKED"
+    ERROR = "ERROR"
+
+    @classmethod
+    def from_ev_charger_data(cls, data: EVChargerData) -> EVChargerState:
+        """Create an `EVChargerState` instance from component data.
+
+        Args:
+            data: ev charger data coming from microgrid.
+
+        Returns:
+            An `EVChargerState` instance.
+        """
+        if data.component_state in (
+            EVChargerComponentState.AUTHORIZATION_REJECTED,
+            EVChargerComponentState.ERROR,
+        ):
+            return EVChargerState.ERROR
+        if data.cable_state == EVChargerCableState.EV_LOCKED:
+            return EVChargerState.EV_LOCKED
+        if data.cable_state == EVChargerCableState.EV_PLUGGED:
+            return EVChargerState.EV_PLUGGED
+        return EVChargerState.IDLE
+
+
+@dataclass(frozen=True)
+class EVChargerPoolStates:
+    """States of all ev chargers in the pool."""
+
+    _states: dict[int, EVChargerState]
+    _changed_component: Optional[int] = None
+
+    def __iter__(self) -> Iterator[tuple[int, EVChargerState]]:
+        """Iterate over states of all ev chargers.
+
+        Returns:
+            An iterator over all ev charger states.
+        """
+        return iter(self._states.items())
+
+    def latest_change(self) -> Optional[tuple[int, EVChargerState]]:
+        """Return the most recent ev charger state change.
+
+        Returns:
+            A tuple with the component ID of an ev charger that just had a state
+                change, and its new state.
+        """
+        if self._changed_component is None:
+            return None
+        return (
+            self._changed_component,
+            self._states.setdefault(
+                self._changed_component, EVChargerState.UNSPECIFIED
+            ),
+        )
+
+
+class StateTracker:
+    """A class for keeping track of the states of all ev chargers in a pool."""
+
+    _states: dict[int, EVChargerState]
+
+    def __init__(self, component_ids: set[int]) -> None:
+        """Create a `_StateTracker` instance.
+
+        Args:
+            component_ids: ev charger component ids to track the states of.
+        """
+        self._component_ids = component_ids
+        self._channel = Broadcast[EVChargerPoolStates](
+            "EVCharger States", resend_latest=True
+        )
```

Comment:
related to https://github.com/frequenz-floss/frequenz-sdk-python/pull/201/files#r1108673277, the type hint might be missing here though

Reply:
Here the typehint is inferred by the convenience thing in python type hints:

``` python
self._channel: Broadcast[int] = Broadcast("name")

## is the same as:

self._channel = Broadcast[int]("name")
```

Reply:
Got it, thanks!